src/Layouts/Partials/LayersPanel.vala

```diff
@@ -137,37 +162,48 @@ public class Akira.Layouts.Partials.LayersPanel : Gtk.ListBox {
         drag_leave.connect (on_drag_leave);
     }
 
-    private void on_drag_data_received (Gdk.DragContext context, int x, int y,
-        Gtk.SelectionData selection_data, uint target_type, uint time) {
-        Akira.Layouts.Partials.Artboard target;
+    private void on_drag_data_received (
+      Gdk.DragContext context,
+      int x, int y,
+      Gtk.SelectionData selection_data,
+      uint target_type, uint time) {
+        Akira.Layouts.Partials.Layer? target;
         Gtk.Widget row;
-        Akira.Layouts.Partials.Artboard source;
+        Akira.Layouts.Partials.Layer? source;
         int new_position;
 
-        target = (Akira.Layouts.Partials.Artboard) get_row_at_y (y);
+        row = ((Gtk.Widget[]) selection_data.get_data ())[0];
+        source = row as Akira.Layouts.Partials.Layer;
+
+        Gtk.Allocation alloc;
+        source.get_allocation (out alloc);
+
+        // In order to determine which position should the dragged
+        // item occupy, we need to check in which gap it is.
+        // By adding half of the height of a row we know between which
+        // rows we want to inser the dragged layer
+        var target_row_y = y + alloc.height / 2;
+
+        target = (Akira.Layouts.Partials.Layer) get_row_at_y (target_row_y);
 
         if (target == null) {
             new_position = -1;
         } else {
             new_position = target.get_index ();
         }
 
-        row = ((Gtk.Widget[]) selection_data.get_data ())[0];
-
-        source = (Akira.Layouts.Partials.Artboard) row.get_ancestor (typeof (Akira.Layouts.Partials.Artboard));
-
         if (source == target) {
             return;
         }
 
-        remove (source);
-        insert (source, new_position);
+        window.event_bus.change_item_z_index (source.model.item, new_position);
+
+        window.main_window.right_sidebar.indicator.visible = false;
```

Comment:
We could use a signal here to clean this up.

Reply:
Do you refer to the action of hiding the `right_sidebar.indicator`? For sure!